src/lib/operators/table_scan/sorted_segment_search.hpp

```diff
@@ -0,0 +1,151 @@
+ public:
+  SortedSegmentSearch(IteratorType begin, IteratorType end, const OrderByMode& order_by,
+                      const PredicateCondition& predicate_condition, const SearchValueType& search_value)
+      : _begin{begin},
+        _end{end},
+        _predicate_condition{predicate_condition},
+        _search_value{search_value},
+        _is_ascending{order_by == OrderByMode::Ascending || order_by == OrderByMode::AscendingNullsLast},
+        _is_nulls_first{order_by == OrderByMode::Ascending || order_by == OrderByMode::Descending} {}
+
+ private:
+  /**
+   * _get_first_bound and _get_last_bound are used to retrieve the lower and upper bound in a sorted segment but are
+   * independent of its sort order. _get_first_bound will always return the bound with the smaller offset and
+   * _get_last_bound will return the bigger offset.
+   * On a segment sorted in ascending order they would work analogously to lower_bound and upper_bound. For descending
+   * sort order _get_first_bound will actually return an upper bound and _get_last_bound the lower one. However, the
+   * first offset will always point to an entry matching the search value, whereas last offset points to the entry
+   * behind the last matching one.
+   */
+  IteratorType _get_first_bound() const {
+    if (_is_ascending) {
+      return std::lower_bound(_begin, _end, _search_value, [](const auto& segment_position, const auto& search_value) {
+        return segment_position.value() < search_value;
+      });
+    } else {
+      return std::lower_bound(_begin, _end, _search_value, [](const auto& segment_position, const auto& search_value) {
+        return segment_position.value() > search_value;
+      });
+    }
+  }
+
+  IteratorType _get_last_bound() const {
+    if (_is_ascending) {
+      return std::upper_bound(_begin, _end, _search_value, [](const auto& search_value, const auto& segment_position) {
+        return segment_position.value() > search_value;
+      });
+    } else {
+      return std::upper_bound(_begin, _end, _search_value, [](const auto& search_value, const auto& segment_position) {
+        return segment_position.value() < search_value;
+      });
+    }
+  }
+
+  // This function sets the offset(s) which delimit the result set based on the predicate condition and the sort order
+  void _set_begin_and_end() {
+    if (_predicate_condition == PredicateCondition::Equals) {
+      _begin = _get_first_bound();
+      _end = _get_last_bound();
+      return;
+    }
+
+    // clang-format off
+    if (_is_ascending) {
+      switch (_predicate_condition) {
+        case PredicateCondition::GreaterThanEquals: _begin = _get_first_bound(); return;
+        case PredicateCondition::GreaterThan: _begin = _get_last_bound(); return;
+        case PredicateCondition::LessThanEquals: _end = _get_last_bound(); return;
+        case PredicateCondition::LessThan: _end = _get_first_bound(); return;
+        default: Fail("Unsupported comparison type encountered");
```

Comment:
we call these "comparison type"s "PredicateConditions" in Hyrise. Let's keep terminology precise.